mlflow/sagemaker/__init__.py

```diff
@@ -354,14 +511,148 @@ def _deploy(role, image_url, app_name, model_s3_path, run_id, region_name):
+
+    eprint("Found active endpoint with arn: {earn}. Updating...".format(earn=endpoint_arn))
+
+    new_model_name = _get_sagemaker_model_name(endpoint_name)
+    new_model_response = sage_client.create_model(
+        ModelName=new_model_name,
+        PrimaryContainer={
+            'ContainerHostname': 'mfs-serve-%s' % new_model_name,
+            'Image': image_url,
+            'ModelDataUrl': model_s3_path,
+            'Environment': {},
+        },
+        ExecutionRoleArn=role,
+        Tags=[{'Key': 'run_id', 'Value': str(run_id)}, ],
+    )
+    eprint("Created new model with arn: %s" % new_model_response["ModelArn"])
+
+    if mode == DEPLOYMENT_MODE_ADD:
+        new_model_weight = 0
+        production_variants = deployed_production_variants
+    
+    elif mode == DEPLOYMENT_MODE_REPLACE:
+        new_model_weight = 1
+        production_variants = [] 
+
+    new_production_variant = {
+                                'VariantName': 'model1',
+                                'ModelName': new_model_name,
+                                'InitialInstanceCount': 1,
+                                'InstanceType': 'ml.m4.xlarge',
+                                'InitialVariantWeight': new_model_weight 
+                             }
+
+    production_variants.append(new_production_variant)
+
+    # Create the new endpoint configuration and update the endpoint
+    # to adopt the new configuration
+    
+    new_config_name = _get_sagemaker_config_name(endpoint_name) 
+    endpoint_config_response = sage_client.create_endpoint_config(
+        EndpointConfigName=new_config_name,
+        ProductionVariants=production_variants,
+        Tags=[
+            {
+                'Key': 'app_name',
+                'Value': endpoint_name,
+            },
+        ],
+    )
+
+    eprint("Created new endpoint configuration with arn: %s" 
+            % endpoint_config_response["EndpointConfigArn"])
+
+    sage_client.update_endpoint(EndpointName=endpoint_name,
+                                EndpointConfigName=new_config_name)
+
+    eprint("Updated endpoint with new configuration!")
+
+    # If applicable, clean up unused models and old configurations
+
+    if not archive:
+        eprint("Cleaning up unused resources...")
+        if mode == DEPLOYMENT_MODE_REPLACE:
+            s3_client = boto3.client('s3')
+            for pv in deployed_production_variants:
+                deployed_model_name = pv["ModelName"]
+                deployed_model_arn = _delete_sagemaker_model(deployed_model_name, 
+                                                             sage_client, 
+                                                             s3_client)
+                eprint("Deleted model with arn: {marn}".format(marn=deployed_model_arn))
+
+        sage_client.delete_endpoint_config(EndpointConfigName=deployed_config_name)
+        eprint("Deleted configuration with arn: {carn}".format(carn=deployed_config_arn))
```

Comment:
nit: "Delete endpoint configuration ..." since we refer to this object as "endpoint configuration" elsewhere (and it is the sagemaker official name)

Reply:
Done